source/common/upstream/upstream_impl.cc

```diff
@@ -831,14 +833,18 @@ void MainPrioritySetImpl::updateCrossPriorityHostMap(const HostVector& hosts_add
   if (mutable_cross_priority_host_map_ == nullptr) {
     // Copy old read only host map to mutable host map.
     mutable_cross_priority_host_map_ = std::make_shared<HostMap>(*const_cross_priority_host_map_);
+    ENVOY_LOG(debug, "cross_priority host map, copying from const, len: {}",
+              const_cross_priority_host_map_->size());
   }
 
   for (const auto& host : hosts_removed) {
     mutable_cross_priority_host_map_->erase(addressToString(host->address()));
+    ENVOY_LOG(debug, "cross_priority host map, removing: {}", addressToString(host->address()));
   }
 
   for (const auto& host : hosts_added) {
     mutable_cross_priority_host_map_->insert({addressToString(host->address()), host});
+    ENVOY_LOG(debug, "cross_priority host map, adding: {}", addressToString(host->address()));
```

Comment:
Based on your problem description, it looks like the problem is that when the host is initially resolved into IP_A, IP_A is inserted into the mutable_cross_priority_host_map_ here. And later on if the host is resolved into IP_B, then IP_B is inserted here. However, the above  (Line 841) mutable_cross_priority_host_map_->erase(addressToString(host->address())) for IP_A is somehow not called.

Intuitively, I would think when the DNS resolving returns a different address, MainPrioritySetImpl::updateCrossPriorityHostMap() will be called with both host_added and host_removed set.  But, it looks like host_removed is empty here. Could you please share which code path is causing that?

Reply:
> Based on your problem description, it looks like the problem is that when the host is initially resolved into IP_A, IP_A is inserted into the mutable_cross_priority_host_map_ here. And later on if the host is resolved into IP_B, then IP_B is inserted here. However, the above (Line 841) mutable_cross_priority_host_map_->erase(addressToString(host->address())) for IP_A is somehow not called.

yep, that's it.

> 
> Intuitively, I would think when the DNS resolving returns a different address, MainPrioritySetImpl::updateCrossPriorityHostMap() will be called with both host_added and host_removed set. But, it looks like host_removed is empty here. Could you please share which code path is causing that?

Nope, in the old implementation, it simply do the priority update on host addition and skip on host subtraction, similar to strict DNS, as @alyssawilk said.
And I think that't the problem.
since it only remove the latest IP in logic DNS when removing an expired host:
https://github.com/envoyproxy/envoy/blob/7a9543d2813e8041a69bba708cdc3c542e0ec8f5/source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc#L268

Reply:
really I don't think having an IP based host map makes sense for DFP hosts, as we can have many "unique" hosts refered to by the same address.  Pinging @wbpcode for thoughts - if it still lived in the LB classes I'd nix it for the DFP cluster but that's not an easy option nowadays.